packages-next/fields-document/ARCHITECTURE.md

```diff
@@ -0,0 +1,53 @@
+
+Besides bespoke normalization for the different types of nodes, there is a set of generic normalization which enforces which block types can have which children, search for `editorSchema` in the to see how this works.
+
+If you want to ensure that the parent of a node is always of a particular type, you must do this by checking the children of a node, not the parent of a node because normalization only runs on nodes that have changed and when a node is moved, normalization will not happen on that node because it did not actually change, normalization will happen on the parent it moved from and to though because both of those node's children array changed.
+
+## User Input
+
+Some behaviour should happen in response to user input, for example markdown shortcuts. These shouldn't happen in normalization because you still want people to be able to write `> ` for example without it becoming a blockquote by undoing it so for these kind of things, they're handled by overriding `editor.insertText`, `editor.insertBreak`, `editor.insertData` or `onKeyDown`(which is a prop to the `Editable` component rather than on the editor instance)
+
+## DocumentFeatures
+
+- `DocumentFeatures` is the configuration of what is enabled in the editor
+- Disabling a feature in `DocumentFeatures` disables it globally and means that it should not be shown in the toolbar and all instances of it should be normalized away
+- Features can also be disabled locally for the current selection(e.g. most things are disabled in code blocks, features can be disabled inside of component child props) which will be represented in the `ToolbarState`
+- The disabled status in the `ToolbarState` refers to the local disabled status only, it doesn't care about the global disabled status so a feature could have `isDisabled: false` locally and be disabled globally
+
+## Component Blocks
+
+Component blocks are the main way that solution developers customise the editor. See the docs for the document field for how they're used. The way they're represented in the structure is that nodes with the type `component-block` store the form data and they have children of type `component-block-prop` and `component-inline-prop` with a `propPath` which points to where in the props structure they are. (note that they are both blocks, inline and block in the name refers to what kind of children they contain)
+
+## Tests
+
+The tests in the document editor are mainly about testing normalization and user input because most of the complexity is there though the UI is also rendered by helpers as a smoke test to ensure nothing immediately blows up. In tests, the document structure is written in jsx, use your editor's autocomplete to see what you can use in the jsx. When making assertions about the document structure of the editor, you should generally use inline snapshots. If you're abstractly testing some behaviour, for example testing markdown shortcuts for each of the marks, use `toEqualEditor`.
```

Comment:
```suggestion
The tests in the document editor are mainly about testing normalization and user input because most of the complexity is there. The UI is also rendered by helpers as a smoke test to ensure nothing immediately blows up. In tests, the document structure is written in jsx, use your editor's autocomplete to see what you can use in the jsx. When making assertions about the document structure of the editor, you should generally use inline snapshots. If you're abstractly testing some behaviour, for example testing markdown shortcuts for each of the marks, use `toEqualEditor`.
```